### PR TITLE
PERF: add method Categorical.__contains__

### PR DESCRIPTION
**Pull request by topper-123**
- [x] closes #21022 
- [x] xref #21369
- [x] tests added / passed
- [x] benchmark added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Currently, membership checks in ``Categorical`` is very slow as explained by @fjetter in #21022. This PR fixes the issue. See also #21369 which fixed the similar issue for ``CategoricalIndex``.

Tests didn't exist beforehand and have been added.

ASV:

```
      before           after         ratio
     [9e982e18]       [28461f0c]
-      4.26±0.7ms         134±20μs     0.03  categoricals.Contains.time_categorical_contains

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

### CODE REVIEW

File: pandas/core/arrays/categorical.py [hunk 1]
```diff
@@ -1847,6 +1847,31 @@ def __iter__(self):
         """Returns an Iterator over the values of this Categorical."""
         return iter(self.get_values().tolist())
 
+    def __contains__(self, key):
+        """Returns True if `key` is in this Categorical."""
+        hash(key)
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: isn’t this very similar to what u did in CI ?
can we unify

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/categorical.py (hunk 1 above)
Comment (topper-123) [reply pandas/core/arrays/categorical.py]: I've made a unified version.

File: pandas/tests/categorical/test_algos.py [hunk 2]
```diff
@@ -71,6 +71,23 @@ def test_isin_empty(empty):
     tm.assert_numpy_array_equal(expected, result)
 
 
+def test_contains():
+
+    c = pd.Categorical(list('aabbca'), categories=list('cab'))
```
Comment (gfyoung) [inline pandas/tests/categorical/test_algos.py]: I know this is not a bug, but can we reference the original issue anyway?

File: pandas/tests/categorical/test_algos.py [hunk 3]
```diff
@@ -71,6 +71,23 @@ def test_isin_empty(empty):
...
+    assert 'b' in c
+    assert 'z' not in c
+    assert np.nan not in c
+
+    # assert codes NOT in index
+    assert 0 not in c
+    assert 1 not in c
+
+    c = pd.Categorical(list('aabbca') + [np.nan], categories=list('cab'))
+
```
Comment (gfyoung) [inline pandas/tests/categorical/test_algos.py]: Nit: unnecessary newline IMO

File: pandas/core/arrays/categorical.py [hunk 4]
```diff
@@ -1847,6 +1847,43 @@ def __iter__(self):
...
         return iter(self.get_values().tolist())
 
+    @staticmethod
+    def _contains(key, categories, container):
+        """Returns True if `key` is in `categories` and the
+        location of `key` in `categories` is in `container`.
+
+        This is a helper method used in :method:`Categorical.__contains__`
+        and in :class:`CategoricalIndex.__contains__`.
+        """
```
Comment (gfyoung) [inline pandas/core/arrays/categorical.py]: As this is not a one-liner doc-string:

* The docstring should start underneath the triple parentheses.
* The first line is a summary should only be one line
* List of parameters and their dtypes
* Return description / type

**Comment by gfyoung** [comment]
@topper-123 : A bunch of nit-picking things, but overall this looks pretty good!

**Comment by topper-123** [comment]
Comments addressed.

File: pandas/core/arrays/categorical.py [hunk 5]
```diff
@@ -1847,6 +1847,66 @@ def __iter__(self):
...
+        This helper method is used in :method:`Categorical.__contains__`
+        and in :class:`CategoricalIndex.__contains__`.
+
+        Returns True if ``key`` is in ``categories`` and the
+        location of ``key`` in ``categories`` is in ``container``.
+
+        Notes
+        -----
+        This method does not check for Nan values. Do that separately
+        before calling this method.
```
Comment (gfyoung) [inline pandas/core/arrays/categorical.py]: Nice!  Couple of things:

* Notes section goes at the bottom
* The first line ("Helper for membership check...") should be underneath the triple quotes.

Comment (topper-123) [reply pandas/core/arrays/categorical.py]: Ok, both done.

But I've never noticed "the first line should be underneath the triple quotes"-rule and it's not widely applied in pandas. Are you sure that's a style rule?

**Review by gfyoung — DISMISSED** [review]
LGTM!

cc @jreback

**Comment by topper-123** [comment]
Ok, won't hold you up to that approval, fixed :-)

**Review by gfyoung — APPROVED** [review]
Appreciate it 😄 .  LGTM!

cc @jreback

File: doc/source/whatsnew/v0.23.2.txt [hunk 6]
```diff
@@ -27,6 +27,8 @@ Performance Improvements
 - Improved performance of membership checks in :class:`CategoricalIndex`
   (i.e. ``x in ci``-style checks are much faster). :meth:`CategoricalIndex.contains`
   is likewise much faster (:issue:`21369`)
+- Improved performance of membership checks in :class:`Categorical`
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.2.txt]: can you make this refernce this PR number (or just coming the what's new into 1 line either way)

File: pandas/core/indexes/category.py [hunk 7]
```diff
@@ -328,23 +327,8 @@ def __contains__(self, key):
         if isna(key):  # if key is a NaN, check if any NaN is in self.
             return self.hasnans
 
-        # is key in self.categories? Then get its location.
```
Comment (jreback) [inline pandas/core/indexes/category.py]: why can't you just do
```return key in self._data``?

File: pandas/core/arrays/categorical.py [hunk 8]
```diff
@@ -1847,6 +1847,67 @@ def __iter__(self):
         """Returns an Iterator over the values of this Categorical."""
         return iter(self.get_values().tolist())
 
+    @staticmethod
+    def _contains(key, categories, container):
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: this seems pretty convoluted, see my comment below

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/category.py (hunk 7 above)
Comment (topper-123) [reply pandas/core/indexes/category.py]: Ok, actually I'v'e got an good idea for this, will reimplement.

File: pandas/core/arrays/categorical.py (hunk 1 above)
Comment (jreback) [reply pandas/core/arrays/categorical.py]: so it looks very similiar. why cannot you just use ``key in self._data`` for the CI

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by topper-123** [comment]
@jreback , using ``._data`` would be slower, as it doesn't use the hashed value in ``._engine``:

```python
>>> n = 100_000
%timeit 'b' in ci._engine
2.6 µs ± 28.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
%timeit 'b' in ci._data
183 µs ± 9.4 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

So there neeeds to be different implementations.

File: pandas/core/arrays/categorical.py [hunk 9]
```diff
@@ -158,6 +158,55 @@ def _maybe_to_categorical(array):
...
+
+    Returns
+    -------
+    is_in : bool
+        True if ``key`` is in ``self.categories`` and location of
+        ``key`` in ``categories`` is in ``container``, else False.
+
+    Notes
+    -----
+    This method does not check for Nan values. Do that separately
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: Nan -> NaN.

File: pandas/core/arrays/categorical.py [hunk 10]
```diff
@@ -1847,6 +1896,15 @@ def __iter__(self):
         """Returns an Iterator over the values of this Categorical."""
         return iter(self.get_values().tolist())
 
+    def __contains__(self, key):
+        """Returns True if `key` is in this Categorical."""
+        hash(key)
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: I think your hash check needs to go into contains itself (null check is ok where it is).

can you add a test on something that isn't hashable as well

File: pandas/core/arrays/categorical.py [hunk 11]
```diff
@@ -1847,6 +1896,15 @@ def __iter__(self):
...
+
+        if isna(key):  # if key is a NaN, check if any NaN is in self.
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: can you put comments on the line above

File: pandas/tests/categorical/test_algos.py [hunk 12]
```diff
@@ -71,6 +71,22 @@ def test_isin_empty(empty):
     tm.assert_numpy_array_equal(expected, result)
 
 
+def test_contains():
```
Comment (jreback) [inline pandas/tests/categorical/test_algos.py]: move to test_indexing, should be other checks already there, de-duplicate with those.

**Review by jreback — CHANGES_REQUESTED** [review]
this ok. some comments. ping when green.

File: pandas/tests/categorical/test_algos.py (hunk 12 above)
Comment (topper-123) [reply pandas/tests/categorical/test_algos.py]: Lookef at test_indexing, didn't find anything very similar. I moved it to test_operators, which is more similar, IMO. Is that ok?

**Review by jorisvandenbossche — COMMENTED** [review]
What is the reason you changed the private  ``._contains`` to public  ``contains``? (it's not needed for the implementation?) 
It's just that I am not sure it is worth adding a new public method.

**Comment by topper-123** [comment]
@jorisvandenbossche, to my understanding these modules are themselves private and it’s not necessary for functions in private modules to be marked as private - that would be superfluous.

When it was a method on Categorical, it *was* needed to mark it private, as Categorical is a public class.

**Comment by topper-123** [comment]
The Travis error is an HTTP error:

```
CondaHTTPError: HTTP 000 CONNECTION FAILED for url <https://repo.anaconda.com/pkgs/main/linux-64/qt-5.9.5-h7e424d6_0.tar.bz2>
Elapsed: -
An HTTP error occurred when trying to retrieve this URL.
HTTP errors are often intermittent, and a simple retry will get you on your way.
real	3m14.046s
user	1m7.620s
sys	0m3.084s
The command "ci/install_travis.sh" failed and exited with 1 during .
Your build has been stopped.
```

so the Travis failure has nothing to do with this PR.

**Comment by jorisvandenbossche** [comment]
> @jorisvandenbossche, to my understanding these modules are themselves private and it’s not necessary for functions in private modules to be marked as private - that would be superfluous.

@topper-123 whoops, sorry, only looked very briefly at the incremental diff and missed that it was a function and not a method.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @topper-123

**Pull request merged**